### PR TITLE
Feat/conflict alert

### LINKED ISSUE #26: Alert if products have conflicts
## Value Statement

As a user, I don't want to use products together that have conflicts, so that I can maximize the benefits from my routine
## Acceptance Criteria

* User needs to be alerted about conflicts when trying to add two products that should not be used together
* Use should be asked whether or not they want to add the product anyway

## Tasks

- [x] Create feature branch
- [x] Write logic to check stored data to see if products have conflict (Check for conflicting idea)
- [x] Render name of the products that are conflicting

## Complexity
large

### PR DESCRIPTION
**Pull request by FriederikeM**
* Alert the user if they are trying to add products that have conflict to the same day & time of day (e.g. Monday morning)

### CODE REVIEW

File: src/components/DailyRoutinePage/DailyRoutine.js [hunk 1]
```diff
@@ -46,9 +47,10 @@ export default function DailyRoutine() {
     return product.days[i].evening;
   });
 
-  function handleEditRoutine(id, name) {
+  function handleEditRoutine(id, name, contraindications) {
```
Comment (mpagels) [inline src/components/DailyRoutinePage/DailyRoutine.js]: What is `contraindications`? I don't understand the word and I don't understand what's in there. In my opinion a better name would be good. Or maybe I'm just stupid to not to see it. 😺 
And try to use camelCase for better readability.

File: src/components/FormModal/FormModal.js [hunk 2]
```diff
@@ -92,27 +98,100 @@ export default function FormModal({ onCancelAdding, id, name }) {
...
 
   function handleMorningClicked(name) {
-    const newMorningChecked = weekRoutine.days.map((day) => {
-      if (day.name === name) {
-        day.morning = !day.morning;
-        return day;
-      } else {
-        return day;
-      }
+    const i = getIndexForWeekday(name);
```
Comment (mpagels) [inline src/components/FormModal/FormModal.js]: I would name the variable `i` to something like `indexOfWeekday` for better readability.

File: src/components/FormModal/FormModal.js [hunk 3]
```diff
@@ -92,27 +98,100 @@ export default function FormModal({ onCancelAdding, id, name }) {
...
+        product.days[i].name === name && product.days[i].morning === true
+    );
+
+    const routineIDs = checkedMornings.map((product) => product.id);
+    const intersection = routineIDs.filter((id) => conflicts.includes(id));
+
+    const conflictName = intersection.map((id) => {
+      const conflictProduct = getProductById(id, products);
+      return conflictProduct.name;
     });
```
Comment (mpagels) [inline src/components/FormModal/FormModal.js]: You could chain this. Save some line of code.
```suggestion
    const conflictName = routineData.filter( 
    (product) => product.days[i].name === name && product.days[i].morning === true)
    .map((product) => product.id)
    .filter((id) => conflicts.includes(id))
    .map((id) => {
      const conflictProduct = getProductById(id, products);
      return conflictProduct.name;
    })
```

File: src/components/FormModal/FormModal.js [hunk 4]
```diff
@@ -92,27 +98,100 @@ export default function FormModal({ onCancelAdding, id, name }) {
...
-    setWeekRoutine({ id: id, days: newMorningChecked, date: weekRoutine.date });
+
+    if (intersection.length > 0) {
```
Comment (mpagels) [inline src/components/FormModal/FormModal.js:118]: I see, that you are using the intersection list for some conditions here. In my change suggestion above I deleted it. May change the changing in a way, that this code is working too. Sorry about that.

File: src/components/FormModal/FormModal.js [hunk 5]
```diff
@@ -92,27 +98,100 @@ export default function FormModal({ onCancelAdding, id, name }) {
...
+          return day;
+        }
+      });
+      setWeekRoutine({
+        id: id,
+        days: newEveningChecked,
+        date: weekRoutine.date,
+      });
+    }
+  }
```
Comment (mpagels) [inline src/components/FormModal/FormModal.js]: This seems to be code you had already written. Maybe you can create a util or service file which you can call to do this stuff, so you don't have to write the code everytime you need the logic.

File: src/components/FormModal/FormModal.js [hunk 6]
```diff
@@ -92,27 +98,100 @@ export default function FormModal({ onCancelAdding, id, name }) {
...
+
+  function handleModalFormSubmit(event) {
+    event.preventDefault();
+
+    const noDaysChecked = weekRoutine.days.map((day) => {
```
Comment (mpagels) [inline src/components/FormModal/FormModal.js]: The variable name is hard to understand. First I thought, in here are the days, which are not checked, but after understanding the code, it seems, that the variable holds a list of booleans. Maybe a good start would be to start with a `is` in the variable word. And maybe end with the word `list`.

```suggestion
    const isNoDaysCheckedList = weekRoutine.days.map((day) => {
```

Read more [here](https://stackoverflow.com/questions/1227998/naming-conventions-what-to-name-a-boolean-variable) about it.

File: src/components/FormModal/FormModal.js [hunk 7]
```diff
@@ -92,27 +98,100 @@ export default function FormModal({ onCancelAdding, id, name }) {
...
+    const nothingSelected = noDaysChecked.every((day) => day === true);
+
+    const checkNotSpecified = weekRoutine.days.map((day) => {
+      return (
+        day.isChecked === true && day.morning === false && day.evening === false
+      );
+    });
+    const noUnspecifiedChecks = checkNotSpecified.every(
+      (check) => check === false
+    );
```
Comment (mpagels) [inline src/components/FormModal/FormModal.js]: Here could  be chaining again a good way to make the code cleaner.

**Review by mpagels — APPROVED** [review]
Very good work Rikki.
Found some small code writing style improvements ;-)

I think it would be very good to refactor the code before starting a new feature. The code is getting very long and complex. I would recommend booking a day to have a look, where do you can simplify code, where is code written two times, and how we can merge this, so you have this only one time. Can you create some sort of `util` or `service` functions to help you to make your code clean.

But after all, excellent job. A lot of complex logic is happening. I like it. Keep pushing.

File: src/components/DailyRoutinePage/DailyRoutine.js [hunk 8]
```diff
@@ -84,7 +92,13 @@ export default function DailyRoutine() {
           packaging={productData.packging}
```
Comment (namirsab) [inline src/components/DailyRoutinePage/DailyRoutine.js:92]: You have `productData` and `product`. I find those names a bit confusing. What is the difference?

File: src/utility/localStorage.js [hunk 9]
```diff
@@ -4,11 +4,16 @@ export function getDataFromLocalStorage() {
...
+    data.push(items);
+    localStorage.setItem("routine", JSON.stringify(data));
+  } else {
+    const index = data.findIndex((routine) => routine.id === items.id);
+    const newData = [...data.slice(0, index), items, ...data.slice(index + 1)];
 
-  localStorage.setItem("routine", JSON.stringify(newData));
+    localStorage.setItem("routine", JSON.stringify(newData));
+  }
 }
```
Comment (namirsab) [inline src/utility/localStorage.js]: I'd create 2 functions, one to add new data, and one to update existing data.
This is a principle (SRP) that explains this, that functions should do one thing and do it well
https://en.wikipedia.org/wiki/Single-responsibility_principle

File: src/components/DailyRoutinePage/DailyRoutine.js (hunk 8 above)
Comment (FriederikeM) [reply src/components/DailyRoutinePage/DailyRoutine.js:92]: @namirsab productData is the information that product holds (name, id, ingredients...) and i get it from my json file
product is one item of the array productsEvening which I get by filtering my local storage data.
So it is both referring to the (same) product. But I am getting the infos from different places

File: src/components/DailyRoutinePage/DailyRoutine.js (hunk 1 above)
Comment (FriederikeM) [reply src/components/DailyRoutinePage/DailyRoutine.js]: Just means that two (or more) things don't go together. I changed it to conflicts to make it easier to understand

File: src/components/FormModal/FormModal.js (hunk 3 above)
Comment (FriederikeM) [reply src/components/FormModal/FormModal.js]: I'll chain the first bit that gets me to the intersection definition and then i will do a separate map for the conflictName, so i can still use the intersection array as a condition for below :)

File: src/components/DailyRoutinePage/DailyRoutine.js (hunk 1 above)
Comment (mpagels) [reply src/components/DailyRoutinePage/DailyRoutine.js]: I looked it up. It is a real word. Today I learned...

File: src/components/FormModal/FormModal.js (hunk 3 above)
Comment (mpagels) [reply src/components/FormModal/FormModal.js]: Very good idea.

**Pull request merged**